File: src/LeetCode/Algorithms/Java/Possible_Bipartition.java
package LeetCode.Algorithms.Java;

// https://leetcode.com/problems/possible-bipartition/
public class Possible_Bipartition {
	public static void main(String[] args) {
		int N = 4;
		int[][] dislikes = {{1, 2}, {1, 3}, {2, 4}};
		
		System.out.println(possibleBipartition(N, dislikes));
	}
	
	public static boolean possibleBipartition(int N, int[][] dislikes) {
		int[][] graph = new int[N + 1][N + 1];
		int[] person = new int[N + 1];
		
		for (int[] dislike : dislikes) {
			graph[dislike[0]][dislike[1]] = 1;
			graph[dislike[1]][dislike[0]] = 1;
		}
		
		for (int i = 1; i <= N; i++) {
			if (person[i] == 0 && !dfs(graph, person, i, 1)) {
				return false;
			}
		}
		
		return true;
	}
	
	public static boolean dfs(int[][] graph, int[] person, int curr, int p) {
		if (person[curr] != 0) {
			return true;
		}
		
		person[curr] = p;
		
		for (int i = 1; i < graph[curr].length; i++) {
			if (graph[curr][i] == 1) {
				if (person[i] == p || !dfs(graph, person, i, -p)) {
					return false;
				}
			}
		}
		
		return true;
	}
}

File: src/LeetCode/Algorithms/JavaScript/Possible_Bipartition.js
/**
 * https://leetcode.com/problems/possible-bipartition/
 * @param {number} N
 * @param {number[][]} dislikes
 * @return {boolean}
 */
var possibleBipartition = function(N, dislikes) {
	let graph = new Array(N + 1);
	let person = new Array(N + 1).fill(-1);
	
	for (let i = 1; i <= N; i++) {
		graph[i] = new Array();
	}
	
	for (let dislike of dislikes) {
		const [x, y] = dislike;
		
		graph[x].push(y);
		graph[y].push(x);
	}
	
	for (let i = 1; i <= N; i++) {
		if (person[i] === -1 && !dfs(graph, person, i, 1)) {
			return false;
		}
	}
	
	return true;
};

const dfs = (graph, person, curr, p) => {
	if (person[curr] !== -1) {
		return person[curr] == p;
	}
	
	person[curr] = p;
	
	for (let i = 0; i < graph[curr].length; i++) {
		if (!dfs(graph, person, graph[curr][i], 1 - p)) {
			return false;
		}
	}
	
	return true;
};

const N = 4;
const dislikes = [[1, 2], [1, 3], [2, 4]];

console.log(possibleBipartition(N, dislikes));